File: src/lira/linguistics/role/phrase_reader.py
"""PhraseReader: spec 12.3's actual `Phrase.read()` implementation --
the only place phrase-level part-of-speech ambiguity gets resolved.
Tries every PhraseType at the given start position (a NOUN_PHRASE and a
VERB_PHRASE can both plausibly start at the same token -- e.g. "state"
is seeded as both NOUN and VERB), ranks every completed SequencePath
with the shared ReadingScorer, and materialises only the winner into
tensor-backed Words (GraphProcessor.materialise_token) -- every other
candidate stays a lightweight Interpretation record (spec 15, 24),
never a second tree of tensor rows for a reading nothing kept."""

from typing import TYPE_CHECKING, List, Optional, Sequence, Tuple

from ..data.clause import Clause
from ..data.interpretation import Interpretation
from ..data.linguistic_unit_kind import LinguisticUnitKind
from ..data.phrase import Phrase
from ..data.phrase_type import PhraseType
from ..data.reading_error import ReadingError, ReadingErrorKind
from ..data.sequencing_obligation import ObligationKind, SequencingObligation
from ..data.token_reading import TokenReading
from ..data.validation_outcome import ValidationOutcome
from .grammar_configurator import GrammarConfigurator, PhraseGrammar
from .graph_processor import GraphProcessor
from .sequence_engine import SequenceEngine, SequencePath, SequenceStep

if TYPE_CHECKING:
    from lira.vocabulary import Word


# Maps an unresolved obligation to the specific ReadingErrorKind spec 21
# names for it; AUXILIARY_REQUIRES_COMPATIBLE_VERB_FORM has no single
# dedicated kind of its own (an auxiliary chain with no following verb
# just fails to produce a valid VERB_PHRASE sequence at all), so it
# falls back to the generic NO_VALID_PHRASE_SEQUENCE.
_OBLIGATION_ERROR_KIND = {
    ObligationKind.DETERMINER_REQUIRES_NOMINAL_HEAD: ReadingErrorKind.INCOMPLETE_DETERMINER_SEQUENCE,
    ObligationKind.PREPOSITION_REQUIRES_OBJECT: ReadingErrorKind.PREPOSITION_MISSING_OBJECT,
    ObligationKind.AUXILIARY_REQUIRES_COMPATIBLE_VERB_FORM: ReadingErrorKind.NO_VALID_PHRASE_SEQUENCE,
    ObligationKind.INFINITIVE_MARKER_REQUIRES_BASE_VERB: ReadingErrorKind.INFINITIVE_MISSING_VERB,
    ObligationKind.CONJUNCTION_REQUIRES_COORDINATED_ELEMENT: ReadingErrorKind.INCOMPLETE_COORDINATION,
}


class PhraseReader:
    def __init__(self, engine: SequenceEngine, graph_processor: GraphProcessor, grammar: GrammarConfigurator):
        self.engine = engine
        self.graph_processor = graph_processor
        self.grammar = grammar

    def read(
        self,
        tokens: Sequence[TokenReading],
        *,
        start_index: int = 0,
        end_index: Optional[int] = None,
        parent_clause: Optional[Clause] = None,
        grammar: Optional[GrammarConfigurator] = None,
    ) -> Phrase:
        active_grammar = grammar or self.grammar
        end_index = len(tokens) if end_index is None else end_index

        if start_index >= end_index:
            return self._unreadable_phrase(tokens, start_index, parent_clause)

        candidates: List[SequencePath] = []
        for phrase_type in PhraseType:
            phrase_grammar = active_grammar.phrase_grammars[phrase_type]
            if phrase_grammar.nested_phrase_after:
                candidates.extend(self._find_prepositional_paths(tokens, start_index, end_index, phrase_grammar))
            else:
                candidates.extend(self.engine.find_valid_sequences(tokens, start_index, phrase_type, end_index=end_index))

        if not candidates:
            return self._unreadable_phrase(tokens, start_index, parent_clause)

        ranked = self.engine.rank_sequences(candidates, tokens)
        winner_key = self.engine.scorer.rank_key(self.engine.scoring_factors(ranked[0], tokens))
        tie_count = sum(
            1 for path in ranked if self.engine.scorer.rank_key(self.engine.scoring_factors(path, tokens)) == winner_key
        )
        alternatives = tuple(
            self._to_interpretation(path, tokens) for path in ranked[1 : 1 + active_grammar.max_alternative_interpretations]
        )
        return self._build_phrase(ranked[0], tokens, active_grammar, parent_clause, alternatives, tie_count)

    # --- PREPOSITIONAL_PHRASE composition (nested_phrase_after) --------
    # find_valid_sequences can't walk this one alone -- see its own
    # docstring -- so PhraseReader composes it directly: confirm the
    # PREPOSITION start, then recurse into the nested NOUN_PHRASE.

    def _find_prepositional_paths(
        self, tokens: Sequence[TokenReading], start_index: int, end_index: int, phrase_grammar: PhraseGrammar,
    ) -> List[SequencePath]:
        token = tokens[start_index]
        if not token.is_known:
            return []
        scope = self.engine.scope_for_phrase_type(phrase_grammar.phrase_type)
        results: List[SequencePath] = []
        for pos in token.candidate_parts_of_speech():
            if pos not in phrase_grammar.start_states:
                continue
            nested_type = self.engine.nested_phrase_for(phrase_grammar, pos)
            if nested_type is None:
                continue
            trigger_step = SequenceStep(token_index=start_index, part_of_speech=pos)
            nested_candidates = self.engine.find_valid_sequences(tokens, start_index + 1, nested_type, end_index=end_index)

            if not nested_candidates:
                # No object at all follows -- the obligation this
                # PREPOSITION raised is never discharged, a definite
                # negative conclusion (spec 20's own worked example is
                # this exact shape: "over" with nothing valid after it).
                open_obligations: Tuple[SequencingObligation, ...] = ()
                if pos in phrase_grammar.obligations_raised:
                    open_obligations = (SequencingObligation(
                        kind=phrase_grammar.obligations_raised[pos], scope=scope, raised_at_index=start_index,
                        description=f"{pos} at token {start_index} requires an object",
                    ),)
                results.append(SequencePath(
                    phrase_type=phrase_grammar.phrase_type, start_index=start_index, end_index=start_index + 1,
                    steps=(trigger_step,), open_obligations=open_obligations,
                ))
                continue

            # An object exists structurally, so PREPOSITION_REQUIRES_OBJECT
            # is discharged regardless of the object's own validity --
            # the object's own UNRESOLVED/INVALID state (unknown tokens,
            # its own open obligations) is what SequencePath.has_unknown_token
            # and validate_sequence's nested_paths check already surface.
            for nested_path in self.engine.rank_sequences(nested_candidates, tokens)[: self.grammar.max_alternative_interpretations]:
                results.append(SequencePath(
                    phrase_type=phrase_grammar.phrase_type, start_index=start_index, end_index=nested_path.end_index,
                    steps=(trigger_step,), open_obligations=(), nested_paths=(nested_path,),
                ))
        return results

    # --- Materialisation --------------------------------------------------

    def _materialise_step(self, token: TokenReading, step: SequenceStep) -> "Word":
        selected = None
        if not step.is_unknown:
            if step.is_marker:
                # A lexical marker (e.g. infinitive "to") is matched by
                # text, not by the POS it was seeded under -- its own
                # top-ranked seeded sense is still the right one to
                # materialise (spec: never invent a POS -- "to" keeps
                # its real seeded PREPOSITION sense as a Word, the
                # phrase's own phrase_type is what records its role
                # here as an infinitive marker, not a relabelled POS).
                selected = token.candidates[0] if token.candidates else None
            else:
                selected = token.identification_for(step.part_of_speech)
        return self.graph_processor.materialise_token(token, token.token_index, selected_candidate=selected)

    def _select_head(self, path: SequencePath, phrase_grammar: PhraseGrammar) -> Optional[SequenceStep]:
        for preferred in phrase_grammar.head_preference:
            for step in reversed(path.steps):
                if not step.is_unknown and not step.is_marker and step.part_of_speech == preferred:
                    return step
        # Nothing in head_preference matched (e.g. "the cat" with "cat"
        # unseeded) -- the wildcard stands in as head so the phrase
        # still has *something* to point to, correctly staying
        # UNRESOLVED rather than headless.
        for step in reversed(path.steps):
            if step.is_unknown:
                return step
        return None

    def _build_phrase(
        self,
        path: SequencePath,
        tokens: Sequence[TokenReading],
        grammar: GrammarConfigurator,
        parent_clause: Optional[Clause],
        alternatives: Tuple[Interpretation, ...],
        tie_count: int,
    ) -> Phrase:
        phrase_grammar = grammar.phrase_grammars[path.phrase_type]
        words = [self._materialise_step(tokens[step.token_index], step) for step in path.steps]
        nested_phrases = [
            self._build_phrase(nested, tokens, grammar, parent_clause, (), 1) for nested in path.nested_paths
        ]
        all_words = words + [w for nested in nested_phrases for w in nested.words]

        head_step = self._select_head(path, phrase_grammar)
        head_word: Optional["Word"] = None
        head_pos = None
        if head_step is not None:
            head_word = words[path.steps.index(head_step)]
            head_pos = head_step.part_of_speech

        selected_pos = tuple(step.part_of_speech for step in path.steps if step.part_of_speech is not None)
        selected_ids = tuple(
            identification
            for step in path.steps
            if step.part_of_speech is not None
            for identification in (tokens[step.token_index].identification_for(step.part_of_speech),)
            if identification is not None
        )

        validation = self.engine.validate_sequence(path)
        factors = self.engine.scoring_factors(path, tokens)
        confidence = self.engine.scorer.confidence(factors, tie_count=tie_count)
        errors = self._build_errors(path, tokens)

        phrase = Phrase(
            text=" ".join(w.text for w in all_words),
            phrase_type=path.phrase_type,
            words=all_words,
            selected_parts_of_speech=selected_pos,
            selected_identifications=selected_ids,
            head_word=head_word,
            head_part_of_speech=head_pos,
            nested_phrases=nested_phrases,
            parent_clause=parent_clause,
            start_position=path.start_index,
            end_position=path.end_index,
            open_obligations=path.open_obligations,
            validation=validation,
            confidence=confidence,
            alternatives=alternatives,
            errors=errors,
        )
        # Only the winning SequencePath ever reaches _build_phrase (the
        # losing candidates stay lightweight Interpretation records via
        # _to_interpretation) -- so, same as materialise_token, this is
        # the one place a Phrase-level tensor row gets allocated (spec
        # 19's containment: every unit gets a system_property).
        phrase.system_property = self.graph_processor.create_property_wrapper(
            phrase, LinguisticUnitKind.Phrase, path.start_index, "PhraseReader_ReadLayer",
        )
        return phrase

    def _build_errors(self, path: SequencePath, tokens: Sequence[TokenReading]) -> Tuple[ReadingError, ...]:
        errors: List[ReadingError] = []
        for step in path.steps:
            if step.is_unknown:
                token = tokens[step.token_index]
                errors.append(ReadingError(
                    kind=ReadingErrorKind.UNKNOWN_VOCABULARY_WORD, level=LinguisticUnitKind.Phrase,
                    message=f'"{token.text}" has no seeded or hydrated part of speech yet',
                    token_index=step.token_index, token_text=token.text,
                ))
        for obligation in path.open_obligations:
            errors.append(ReadingError(
                kind=_OBLIGATION_ERROR_KIND.get(obligation.kind, ReadingErrorKind.NO_VALID_PHRASE_SEQUENCE),
                level=LinguisticUnitKind.Phrase, message=obligation.description,
                token_index=obligation.raised_at_index, open_scope=obligation.scope,
                unfinished_obligation=obligation.kind,
            ))
        # Errors from a nested phrase (e.g. the NOUN_PHRASE nested under
        # a PREPOSITIONAL_PHRASE) live on that nested Phrase object
        # itself (built separately in _build_phrase's own recursive
        # call) -- not duplicated here.
        return tuple(errors)

    def _to_interpretation(self, path: SequencePath, tokens: Sequence[TokenReading]) -> Interpretation:
        factors = self.engine.scoring_factors(path, tokens)
        return Interpretation(
            selected_parts_of_speech=tuple(step.part_of_speech for step in path.steps if step.part_of_speech is not None),
            selected_entry_ids=tuple(None for _ in path.steps),
            phrase_spans=((path.phrase_type, path.start_index, path.end_index),),
            clause_spans=(),
            open_obligations=path.open_obligations,
            validation=self.engine.validate_sequence(path),
            confidence=self.engine.scorer.confidence(factors),
            rank_key=self.engine.scorer.rank_key(factors),
        )

    def _unreadable_phrase(
        self, tokens: Sequence[TokenReading], start_index: int, parent_clause: Optional[Clause],
    ) -> Phrase:
        token = tokens[start_index] if start_index < len(tokens) else None
        error = ReadingError(
            kind=ReadingErrorKind.NO_VALID_PHRASE_SEQUENCE, level=LinguisticUnitKind.Phrase,
            message="No phrase grammar accepts a token here",
            token_index=start_index if token is not None else None,
            token_text=token.text if token is not None else None,
            seeded_candidate_parts_of_speech=token.candidate_parts_of_speech() if token is not None else (),
        )
        phrase = Phrase(
            text=token.text if token is not None else "",
            phrase_type=None, words=[], parent_clause=parent_clause,
            start_position=start_index, end_position=start_index + (1 if token is not None else 0),
            validation=ValidationOutcome.INVALID, confidence=0.0, errors=(error,),
        )
        phrase.system_property = self.graph_processor.create_property_wrapper(
            phrase, LinguisticUnitKind.Phrase, start_index, "PhraseReader_ReadLayer",
        )
        return phrase

File: prototype/src/lira/linguistics/role/phrase_reader.ts
import { PartOfSpeech } from "../../vocabulary/data/part_of_speech";
import type { Word } from "../../vocabulary/data/word";
import type { Clause } from "../data/clause";
import type { Interpretation } from "../data/interpretation";
import { LinguisticUnitKind } from "../data/linguistic_unit_kind";
import { createPhrase, type Phrase } from "../data/phrase";
import { PhraseType } from "../data/phrase_type";
import { createReadingError, ReadingErrorKind, type ReadingError } from "../data/reading_error";
import { ObligationKind, type SequencingObligation } from "../data/sequencing_obligation";
import { candidatePartsOfSpeech, identificationFor, isKnown, type TokenReading } from "../data/token_reading";
import { ValidationOutcome } from "../data/validation_outcome";
import type { GrammarConfigurator, PhraseGrammar } from "./grammar_configurator";
import type { GraphProcessor } from "./graph_processor";
import { createSequenceStep, type SequenceEngine, type SequencePath, type SequenceStep } from "./sequence_engine";

/** PhraseReader: spec 12.3's actual readPhrase() implementation -- the
 * only place phrase-level part-of-speech ambiguity gets resolved.
 * Tries every PhraseType at the given start position (a NOUN_PHRASE and
 * a VERB_PHRASE can both plausibly start at the same token -- e.g.
 * "state" is seeded as both NOUN and VERB), ranks every completed
 * SequencePath with the shared ReadingScorer, and materialises only the
 * winner into tensor-backed Words (GraphProcessor.materialiseToken) --
 * every other candidate stays a lightweight Interpretation record (spec
 * 15, 24), never a second tree of tensor rows for a reading nothing
 * kept.
 *
 * Ported from linguistics/role/phrase_reader.py. */

// Maps an unresolved obligation to the specific ReadingErrorKind spec 21
// names for it; AUXILIARY_REQUIRES_COMPATIBLE_VERB_FORM has no single
// dedicated kind of its own (an auxiliary chain with no following verb
// just fails to produce a valid VERB_PHRASE sequence at all), so it
// falls back to the generic NO_VALID_PHRASE_SEQUENCE.
const OBLIGATION_ERROR_KIND: Partial<Record<ObligationKind, ReadingErrorKind>> = {
  [ObligationKind.DETERMINER_REQUIRES_NOMINAL_HEAD]: ReadingErrorKind.INCOMPLETE_DETERMINER_SEQUENCE,
  [ObligationKind.PREPOSITION_REQUIRES_OBJECT]: ReadingErrorKind.PREPOSITION_MISSING_OBJECT,
  [ObligationKind.AUXILIARY_REQUIRES_COMPATIBLE_VERB_FORM]: ReadingErrorKind.NO_VALID_PHRASE_SEQUENCE,
  [ObligationKind.INFINITIVE_MARKER_REQUIRES_BASE_VERB]: ReadingErrorKind.INFINITIVE_MISSING_VERB,
  [ObligationKind.CONJUNCTION_REQUIRES_COORDINATED_ELEMENT]: ReadingErrorKind.INCOMPLETE_COORDINATION,
};

export interface PhraseReadOptions {
  startIndex?: number;
  endIndex?: number;
  parentClause?: Clause;
  grammar?: GrammarConfigurator;
  trace?: unknown[];
}

export class PhraseReader {
  constructor(
    public readonly engine: SequenceEngine,
    public readonly graphProcessor: GraphProcessor,
    private readonly grammar: GrammarConfigurator,
  ) {}

  /** `trace`, when an array is passed, gets one JSON-safe record
   * pushed per call describing *every* PhraseType attempted at this
   * start position -- not just the winner. Purely observational:
   * passing `trace` unset is byte-for-byte the original behaviour,
   * since nothing about the search or ranking changes, only what gets
   * recorded about it. */
  read(tokens: readonly TokenReading[], options: PhraseReadOptions = {}): Phrase {
    const activeGrammar = options.grammar ?? this.grammar;
    const startIndex = options.startIndex ?? 0;
    const endIndex = options.endIndex ?? tokens.length;
    const { parentClause, trace } = options;

    if (startIndex >= endIndex) {
      const phrase = this.unreadablePhrase(tokens, startIndex, parentClause);
      trace?.push(this.positionTrace(tokens, startIndex, activeGrammar, new Map(), undefined, phrase));
      return phrase;
    }

    const perTypeCandidates = new Map<PhraseType, SequencePath[]>();
    const candidates: SequencePath[] = [];
    for (const phraseType of Object.values(PhraseType).filter((v): v is PhraseType => typeof v === "number")) {
      const phraseGrammar = activeGrammar.phraseGrammars.get(phraseType);
      if (!phraseGrammar) continue;
      const found = phraseGrammar.nestedPhraseAfter.size > 0
        ? this.findPrepositionalPaths(tokens, startIndex, endIndex, phraseGrammar)
        : this.engine.findValidSequences(tokens, startIndex, phraseType, endIndex);
      perTypeCandidates.set(phraseType, [...found]);
      candidates.push(...found);
    }

    if (candidates.length === 0) {
      const phrase = this.unreadablePhrase(tokens, startIndex, parentClause);
      trace?.push(this.positionTrace(tokens, startIndex, activeGrammar, perTypeCandidates, undefined, phrase));
      return phrase;
    }

    const ranked = this.engine.rankSequences(candidates, tokens);
    const winningPath = ranked[0];
    const winnerKey = this.engine.scorer.rankKey(this.engine.scoringFactors(winningPath, tokens));
    const tieCount = ranked.filter((path) => keysEqual(this.engine.scorer.rankKey(this.engine.scoringFactors(path, tokens)), winnerKey)).length;
    const alternatives = ranked.slice(1, 1 + activeGrammar.maxAlternativeInterpretations).map((path) => this.toInterpretation(path, tokens));
    const phrase = this.buildPhrase(winningPath, tokens, activeGrammar, parentClause, alternatives, tieCount);
    trace?.push(this.positionTrace(tokens, startIndex, activeGrammar, perTypeCandidates, winningPath, phrase));
    return phrase;
  }

  // --- PREPOSITIONAL_PHRASE composition (nestedPhraseAfter) --------
  // findValidSequences can't walk this one alone -- see its own
  // docstring -- so PhraseReader composes it directly: confirm the
  // PREPOSITION start, then recurse into the nested NOUN_PHRASE.

  private findPrepositionalPaths(
    tokens: readonly TokenReading[], startIndex: number, endIndex: number, phraseGrammar: PhraseGrammar,
  ): SequencePath[] {
    const token = tokens[startIndex];
    if (!isKnown(token)) return [];
    const scope = this.engine.scopeForPhraseType(phraseGrammar.phraseType);
    const results: SequencePath[] = [];
    for (const pos of candidatePartsOfSpeech(token)) {
      if (!phraseGrammar.startStates.has(pos)) continue;
      const nestedType = this.engine.nestedPhraseFor(phraseGrammar, pos);
      if (nestedType === undefined) continue;
      const triggerStep = createSequenceStep(startIndex, pos);
      const nestedCandidates = this.engine.findValidSequences(tokens, startIndex + 1, nestedType, endIndex);

      if (nestedCandidates.length === 0) {
        // No object at all follows -- the obligation this PREPOSITION
        // raised is never discharged, a definite negative conclusion.
        let openObligations: readonly SequencingObligation[] = [];
        const raisedKind = phraseGrammar.obligationsRaised.get(pos);
        if (raisedKind !== undefined) {
          openObligations = [{
            kind: raisedKind, scope, raisedAtIndex: startIndex,
            description: `POS ${pos} at token ${startIndex} requires an object`,
          }];
        }
        results.push({
          phraseType: phraseGrammar.phraseType, startIndex, endIndex: startIndex + 1,
          steps: [triggerStep], openObligations, nestedPaths: [],
        });
        continue;
      }

      // An object exists structurally, so PREPOSITION_REQUIRES_OBJECT is
      // discharged regardless of the object's own validity -- the
      // object's own UNRESOLVED/INVALID state is what hasUnknownToken
      // and validateSequence's nestedPaths check already surface.
      for (const nestedPath of this.engine.rankSequences(nestedCandidates, tokens).slice(0, this.grammar.maxAlternativeInterpretations)) {
        results.push({
          phraseType: phraseGrammar.phraseType, startIndex, endIndex: nestedPath.endIndex,
          steps: [triggerStep], openObligations: [], nestedPaths: [nestedPath],
        });
      }
    }
    return results;
  }

  // --- Materialisation --------------------------------------------------

  private materialiseStep(token: TokenReading, step: SequenceStep): Word {
    let selected;
    if (!step.isUnknown) {
      if (step.isMarker) {
        // A lexical marker (e.g. infinitive "to") is matched by text,
        // not by the POS it was seeded under -- its own top-ranked
        // seeded sense is still the right one to materialise. The
        // phrase's own phraseType is what records its role here as an
        // infinitive marker, not a relabelled POS.
        selected = token.candidates.length > 0 ? token.candidates[0] : undefined;
      } else if (step.partOfSpeech !== undefined) {
        selected = identificationFor(token, step.partOfSpeech);
      }
    }
    return this.graphProcessor.materialiseToken(token, token.tokenIndex, selected);
  }

  private selectHead(path: SequencePath, phraseGrammar: PhraseGrammar): SequenceStep | undefined {
    for (const preferred of phraseGrammar.headPreference) {
      for (let i = path.steps.length - 1; i >= 0; i--) {
        const step = path.steps[i];
        if (!step.isUnknown && !step.isMarker && step.partOfSpeech === preferred) return step;
      }
    }
    // Nothing in headPreference matched (e.g. "the cat" with "cat"
    // unseeded) -- the wildcard stands in as head so the phrase still
    // has *something* to point to, correctly staying UNRESOLVED rather
    // than headless.
    for (let i = path.steps.length - 1; i >= 0; i--) {
      if (path.steps[i].isUnknown) return path.steps[i];
    }
    return undefined;
  }

  private buildPhrase(
    path: SequencePath, tokens: readonly TokenReading[], grammar: GrammarConfigurator,
    parentClause: Clause | undefined, alternatives: readonly Interpretation[], tieCount: number,
  ): Phrase {
    const phraseGrammar = grammar.phraseGrammars.get(path.phraseType) as PhraseGrammar;
    const words = path.steps.map((step) => this.materialiseStep(tokens[step.tokenIndex], step));
    const nestedPhrases = path.nestedPaths.map((nested) => this.buildPhrase(nested, tokens, grammar, parentClause, [], 1));
    const allWords = [...words, ...nestedPhrases.flatMap((nested) => nested.words)];

    const headStep = this.selectHead(path, phraseGrammar);
    let headWord: Word | undefined;
    let headPos;
    if (headStep !== undefined) {
      headWord = words[path.steps.indexOf(headStep)];
      headPos = headStep.partOfSpeech;
    }

    const selectedPos = path.steps.map((step) => step.partOfSpeech).filter((pos): pos is NonNullable<typeof pos> => pos !== undefined);
    const selectedIds = path.steps
      .filter((step) => step.partOfSpeech !== undefined)
      .map((step) => identificationFor(tokens[step.tokenIndex], step.partOfSpeech!))
      .filter((identification) => identification !== undefined);

    const validation = this.engine.validateSequence(path);
    const factors = this.engine.scoringFactors(path, tokens);
    const confidence = this.engine.scorer.confidence(factors, tieCount);
    const errors = this.buildErrors(path, tokens);

    const phrase = createPhrase({
      text: allWords.map((w) => w.text).join(" "),
      phraseType: path.phraseType,
      words: allWords,
      selectedPartsOfSpeech: selectedPos,
      selectedIdentifications: selectedIds,
      headWord, headPartOfSpeech: headPos,
      nestedPhrases,
      parentClause,
      startPosition: path.startIndex, endPosition: path.endIndex,
      openObligations: path.openObligations,
      validation, confidence, alternatives, errors,
    });
    // Only the winning SequencePath ever reaches buildPhrase (the
    // losing candidates stay lightweight Interpretation records via
    // toInterpretation) -- so this is the one place a Phrase-level
    // tensor row gets allocated (spec 19's containment: every unit
    // gets a systemProperty).
    phrase.systemProperty = this.graphProcessor.createPropertyWrapper(
      phrase, LinguisticUnitKind.Phrase, path.startIndex, "PhraseReader_ReadLayer",
    );
    return phrase;
  }

  private buildErrors(path: SequencePath, tokens: readonly TokenReading[]): readonly ReadingError[] {
    const errors: ReadingError[] = [];
    for (const step of path.steps) {
      if (step.isUnknown) {
        const token = tokens[step.tokenIndex];
        errors.push(createReadingError({
          kind: ReadingErrorKind.UNKNOWN_VOCABULARY_WORD, level: LinguisticUnitKind.Phrase,
          message: `"${token.text}" has no seeded or hydrated part of speech yet`,
          tokenIndex: step.tokenIndex, tokenText: token.text,
        }));
      }
    }
    for (const obligation of path.openObligations) {
      errors.push(createReadingError({
        kind: OBLIGATION_ERROR_KIND[obligation.kind] ?? ReadingErrorKind.NO_VALID_PHRASE_SEQUENCE,
        level: LinguisticUnitKind.Phrase, message: obligation.description,
        tokenIndex: obligation.raisedAtIndex, openScope: obligation.scope,
        unfinishedObligation: obligation.kind,
      }));
    }
    // Errors from a nested phrase (e.g. the NOUN_PHRASE nested under a
    // PREPOSITIONAL_PHRASE) live on that nested Phrase object itself
    // (built separately in buildPhrase's own recursive call) -- not
    // duplicated here.
    return errors;
  }

  private toInterpretation(path: SequencePath, tokens: readonly TokenReading[]): Interpretation {
    const factors = this.engine.scoringFactors(path, tokens);
    return {
      selectedPartsOfSpeech: path.steps.map((step) => step.partOfSpeech).filter((pos): pos is NonNullable<typeof pos> => pos !== undefined),
      selectedEntryIds: path.steps.map(() => undefined),
      phraseSpans: [[path.phraseType, path.startIndex, path.endIndex]],
      clauseSpans: [],
      openObligations: path.openObligations,
      validation: this.engine.validateSequence(path),
      confidence: this.engine.scorer.confidence(factors),
      rankKey: this.engine.scorer.rankKey(factors),
      errors: [],
    };
  }

  // --- Trace (sentence reader UI's "full trace" panel; kept as real
  // Service output even though the UI itself isn't ported here) ------

  private pathText(path: SequencePath, tokens: readonly TokenReading[]): string {
    const words = path.steps.map((step) => tokens[step.tokenIndex].text);
    for (const nested of path.nestedPaths) words.push(this.pathText(nested, tokens));
    return words.join(" ");
  }

  private positionTrace(
    tokens: readonly TokenReading[], startIndex: number, grammar: GrammarConfigurator,
    perTypeCandidates: Map<PhraseType, SequencePath[]>, winningPath: SequencePath | undefined, winningPhrase: Phrase,
  ): unknown {
    const token = startIndex < tokens.length ? tokens[startIndex] : undefined;
    const attempts = [];

    for (const [phraseType, phraseGrammar] of grammar.phraseGrammars) {
      const paths = perTypeCandidates.get(phraseType) ?? [];

      let requiredStart: string[];
      let startMatch: boolean;
      if (phraseGrammar.markerForms.size > 0) {
        requiredStart = [...phraseGrammar.markerForms].sort();
        startMatch = token !== undefined && phraseGrammar.markerForms.has(token.text.toLowerCase());
      } else {
        requiredStart = [...phraseGrammar.startStates].map((pos) => PartOfSpeech[pos]).sort();
        startMatch = token !== undefined && isKnown(token) && candidatePartsOfSpeech(token).some((pos) => phraseGrammar.startStates.has(pos));
      }

      const completions = paths.map((path) => ({
        text: this.pathText(path, tokens),
        endIndex: path.endIndex,
        validation: ValidationOutcome[this.engine.validateSequence(path)],
        confidence: Math.round(this.engine.scorer.confidence(this.engine.scoringFactors(path, tokens)) * 10000) / 10000,
        isWinner: path === winningPath,
      }));

      let rejectionReason: string | null = null;
      if (paths.length > 0) {
        rejectionReason = null;
      } else if (token === undefined) {
        rejectionReason = "no token at this position";
      } else if (phraseGrammar.markerForms.size > 0) {
        rejectionReason = `token text does not match the required marker form(s) ${JSON.stringify(requiredStart)}`;
      } else if (!isKnown(token)) {
        rejectionReason = `"${token.text}" is unseeded -- no candidate part of speech to match a required start state`;
      } else if (!startMatch) {
        const seeded = candidatePartsOfSpeech(token).map((pos) => PartOfSpeech[pos]);
        rejectionReason = `token's seeded part(s) of speech ${JSON.stringify(seeded)} don't include a required start ${JSON.stringify(requiredStart)}`;
      } else {
        rejectionReason = "start matched, but no valid completion could be found from here";
      }

      attempts.push({
        phraseType: PhraseType[phraseType],
        requiredStart, startMatch, completions, rejectionReason,
      });
    }

    return {
      startIndex,
      tokenText: token?.text ?? null,
      candidatePartsOfSpeech: token !== undefined ? candidatePartsOfSpeech(token).map((pos) => PartOfSpeech[pos]) : [],
      isKnown: token !== undefined ? isKnown(token) : null,
      attempts,
      winnerPhraseType: winningPhrase.phraseType !== undefined ? PhraseType[winningPhrase.phraseType] : null,
      winnerText: winningPhrase.text,
      winnerValidation: ValidationOutcome[winningPhrase.validation],
      winnerEndIndex: winningPhrase.endPosition,
    };
  }

  private unreadablePhrase(tokens: readonly TokenReading[], startIndex: number, parentClause: Clause | undefined): Phrase {
    const token = startIndex < tokens.length ? tokens[startIndex] : undefined;
    const error = createReadingError({
      kind: ReadingErrorKind.NO_VALID_PHRASE_SEQUENCE, level: LinguisticUnitKind.Phrase,
      message: "No phrase grammar accepts a token here",
      tokenIndex: token !== undefined ? startIndex : undefined,
      tokenText: token?.text,
      seededCandidatePartsOfSpeech: token !== undefined ? candidatePartsOfSpeech(token) : [],
    });
    const phrase = createPhrase({
      text: token?.text ?? "",
      phraseType: undefined, words: [], parentClause,
      startPosition: startIndex, endPosition: startIndex + (token !== undefined ? 1 : 0),
      validation: ValidationOutcome.INVALID, confidence: 0.0, errors: [error],
    });
    phrase.systemProperty = this.graphProcessor.createPropertyWrapper(
      phrase, LinguisticUnitKind.Phrase, startIndex, "PhraseReader_ReadLayer",
    );
    return phrase;
  }
}

function keysEqual(a: readonly number[], b: readonly number[]): boolean {
  return a.length === b.length && a.every((value, index) => value === b[index]);
}
